tfjs-backend-webgpu/src/kernels/from_pixels_webgpu.ts

```diff
@@ -26,35 +26,132 @@ export class FromPixelsProgram implements WebGPUProgram {
   variableNames = ['A'];
   dispatchLayout: {x: number[]};
   dispatch: [number, number, number];
+  needsShapesUniforms = false;
+  pipeline: GPUComputePipeline;
+  bindGroupLayout: GPUBindGroupLayout;
+
+  uniform: GPUBuffer;
+  lastUniformContent = [0, 0, 0];
+
+  inputTexture: GPUTexture = null;
+  lastPixelSize = {width: 0, height: 0};
 
   constructor(outputShape: number[]) {
     this.outputShape = outputShape;
     this.dispatchLayout = flatDispatchLayout(this.outputShape);
     this.dispatch = computeDispatch(this.dispatchLayout, this.outputShape);
-
+    const workGroupSize = 256;
     this.userCode = `
-      void main() {
-        ivec3 coords = getOutputCoords();
-        int texR = coords[0];
-        int texC = coords[1];
-        int depth = coords[2];
-        vec2 uv = (vec2(texC, texR) + halfCR) / vec2(outShape.yx);
-
-        vec4 values = texelFetch(A, uv);
-        float value;
-        if (depth == 0) {
-          value = values.r;
-        } else if (depth == 1) {
-          value = values.g;
-        } else if (depth == 2) {
-          value = values.b;
-        } else if (depth == 3) {
-          value = values.a;
-        }
+    layout (local_size_x = ${workGroupSize},
+      local_size_y = 1,
+      local_size_z = 1) in;
+    layout(set = 0, binding = 1, rgba8) uniform readonly image2D srcImage;
+    layout(set = 0, binding = 2) uniform Meta {
+      int size;
+      int width;
+      int numChannels;} outShape;
+
+    ivec3 getOutputCoordsIB() {
+      int index0 = int(gl_GlobalInvocationID[0]) * outShape.numChannels;
+      int d0 = index0 / (outShape.numChannels * outShape.width);
+      index0 -= d0 * (outShape.numChannels * outShape.width);
+      int d1 = index0 / outShape.numChannels;
+      int d2 = index0 - d1 * outShape.numChannels;
+      return ivec3(d0,d1,d2);
+    }
 
-        setOutput(floor(value * 255.0 + 0.5));
+    void main() {
+      ivec3 coords = getOutputCoordsIB();
```

Comment:
Maybe you can directly use `getCoordsFromFlatIndex(int(gl_GlobalInvocationID.x) * outShape.numChannels)` function in shader_preprocessor.ts and don't need to define a new one.

Reply:
Thanks for the suggestion. I can have a try.

Reply:
It works fine! Done.